doc/src/manual/methods.md

```diff
@@ -515,6 +515,226 @@ julia> wait(schedule(t, 1))
 "definition for Int"
 ```
 
+## Design Patterns with Parametric Methods
+
+
+While complex dispatch logic is not required for performance or usability,
+sometimes it can be the best way to express some algorithm.
+Here are a few common design patterns that come up sometimes when using dispatch in this way.
+
+### Extracting the type parameter from a super-type
+
+
+Here is the correct code template for returning the element-type `T`
+of any arbitrary subtype of `AbstractArray`:
+
+```julia
+abstract AbstractArray{T, N}
+eltype(::Type{<:AbstractArray{T}}) where {T} = T
+```
+using so-called triangular dispatch.
+
+Another way, which used to be the only correct way before the advent of
+triangular dispatch, is:
+
+```julia
+abstract AbstractArray{T, N}
+eltype(::Type{AbstractArray}) = Any
+eltype{T}(::Type{AbstractArray{T}}) = T
+eltype{T, N}(::Type{AbstractArray{T, N}}) = T
+eltype{A<:AbstractArray}(::Type{A}) = eltype(supertype(A))
+```
```

Comment:
Ok, so I looked into this a bit more.  That error you report also happens to `eltype`:
```
julia> eltype(Array{Array{T, 1}, 1} where T)                                                                                                          
ERROR: UndefVarError: E not defined                                                                                                                   
Stacktrace:                                                                                                                                           
 [1] eltype(::Type{Array{Array{T,1},1} where T}) at ./abstractarray.jl:106                                                                            
```
but this is a bug in Julia, right? Or how can it be that the method matches but `T` is not defined:
```
julia> ff(::Type{<:AbstractArray{T}}) where {T} = T                                                                                                   
ff (generic function with 1 method)                                                                                                                   

julia> ff(Array{Array{T, 1}, 1} where T)                                                                                                              
ERROR: UndefVarError: T not defined                                                                                                                   
Stacktrace:                                                                                                                                           
 [1] ff(::Type{Array{Array{T,1},1} where T}) at ./REPL[3]:1                                                                                           
```
adding `@isdefined(T) ? T : Any` to `ff` does indeed fix it.

Anyway, as your methods and my two (and `eltype` itself) seem to be afflicted by this, I still think we should include the simpler version.

Aside, note that your methods do not work on 0.6 at all (presumably because of 0.6 bugs).  Here the gist for running all tests:
https://gist.github.com/mauro3/856423b2f746de67639314d68bec5e50